File: Coding-Interviews/18-删除链表的节点.cpp
#include <memory>
using namespace std;
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
    ListNode* deleteNode(ListNode* head, int val)
    {
        auto dummy  = make_unique<ListNode>(0);
        dummy->next = head;
        auto pre    = dummy.get();
        auto cur    = head;
        while (cur) {
            if (cur->val == val)
                break;
            pre = cur;
            cur = cur->next;
        }
        pre->next = cur->next;
        return dummy->next;
    }
};

File: Coding-Interviews/18-删除链表的节点.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    def deleteNode(self, head: ListNode, val: int) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head
        pre = dummy
        cur = head
        while cur:
            if cur.val == val:
                pre.next = cur.next
                return dummy.next
            pre = cur
            cur = cur.next
        return dummy.next
